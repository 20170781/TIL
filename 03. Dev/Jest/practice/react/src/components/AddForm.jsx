import React, { memo } from 'react';

const AddForm = memo((props) => {
  const inputRef = React.createRef();

  const onSubmit = (event) => {
    event.preventDefault();
    const name = inputRef.current.value;
    name && props.onAdd(name);
    inputRef.current.value = '';
  };

  return (
    <form className="add-form" onSubmit={onSubmit}>
      <input
        ref={inputRef}
        type="text"
        className="add-input"
        placeholder="To-do"
      />
      <button className="add-button">Add</button>
    </form>
  );
});

export default AddForm;
