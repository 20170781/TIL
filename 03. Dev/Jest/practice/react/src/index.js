import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './app/app';
import '@fortawesome/fontawesome-free/js/all.js';
import TodoPresenter from './logics/todo_presenter';

const todoPresenter = new TodoPresenter([
  { id: 1, name: 'JavaScript', count: 0 },
  { id: 2, name: 'TypeScript', count: 0 },
  { id: 3, name: 'React', count: 0 },
]);

ReactDOM.render(
  <React.StrictMode>
    <App presenter={todoPresenter} />
  </React.StrictMode>,
  document.getElementById('root')
);
