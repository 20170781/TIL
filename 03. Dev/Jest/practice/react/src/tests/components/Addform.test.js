import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AddForm from '../../components/AddForm';

describe('AddForm', () => {
  test('snapshot', () => {
    const component = render(<AddForm onAdd={jest.fn()} />);

    expect(component.container).toMatchSnapshot();
  });

  describe('Form submit', () => {
    let onAdd;
    let inputComp;
    let buttonComp;

    beforeEach(() => {
      onAdd = jest.fn();

      render(<AddForm onAdd={onAdd} />); // 컴포넌트 렌더링
      inputComp = screen.getByPlaceholderText('Todo'); // input 태그
      buttonComp = screen.getByText('Add'); // button 테그
    });

    test('calls onAdd when button is clicked and valid Todo Entered', () => {
      const newTodo = 'new todo';
      userEvent.type(inputComp, newTodo); // fireEvent도 가능
      userEvent.click(buttonComp);

      expect(onAdd).toHaveBeenCalledWith(newTodo);
      expect(onAdd).toHaveBeenCalledTimes(1);
    });

    test('does not call onAdd when the Todo is empty', () => {
      userEvent.type(inputComp, ''); // fireEvent도 가능
      userEvent.click(buttonComp);

      expect(onAdd).toHaveBeenCalledTimes(0);
    });
  });
});
