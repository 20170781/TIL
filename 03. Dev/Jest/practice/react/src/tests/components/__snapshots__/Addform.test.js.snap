// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddForm snapshot 1`] = `
<div>
  <form
    class="add-form"
  >
    <input
      class="add-input"
      placeholder="Todo"
      type="text"
    />
    <button
      class="add-button"
    >
      Add
    </button>
  </form>
</div>
`;
